cpp/src/replace/replace.cu

```diff
@@ -19,130 +19,324 @@
-                      const T*                    d_new_values)
+  void replace_kernel(const T* __restrict__           input_data,
+                      bit_mask_t const * __restrict__ input_valid,
+                      T * __restrict__          output_data,
+                      bit_mask_t * __restrict__ output_valid,
+                      gdf_size_type * __restrict__    output_valid_count,
+                      gdf_size_type                   nrows,
+                      const T* __restrict__ values_to_replace_begin,
+                      const T* __restrict__ values_to_replace_end,
+                      const T* __restrict__           d_replacement_values,
+                      bit_mask_t const * __restrict__ replacement_valid)
   {
-    gdf_size_type i = blockIdx.x * blockDim.x + threadIdx.x;
-    while(i < nrows)
-    {
-      auto found_ptr = thrust::find(thrust::seq, old_values_begin, old_values_end, d_col_data[i]);
+  gdf_size_type i = blockIdx.x * blockDim.x + threadIdx.x;
+
+  uint32_t active_mask = 0xffffffff;
+  active_mask = __ballot_sync(active_mask, i < nrows);
+
+  while (i < nrows) {
+    bool output_is_valid = true;
+    uint32_t bitmask = 0xffffffff;
 
-      if (found_ptr != old_values_end) {
-          auto d = thrust::distance(old_values_begin, found_ptr);
-          d_col_data[i] = d_new_values[d];
+    if (input_has_nulls) {
+      bool const input_is_valid{bit_mask::is_valid(input_valid, i)};
+      output_is_valid = input_is_valid;
+
+      bitmask = __ballot_sync(active_mask, input_is_valid);
+
+      if (input_is_valid) {
+        thrust::tie(output_data[i], output_is_valid)  =
+            get_new_value<T, replacement_has_nulls>(i, input_data,
+                                      values_to_replace_begin,
+                                      values_to_replace_end,
+                                      d_replacement_values,
+                                      replacement_valid);
       }
 
-      i += blockDim.x * gridDim.x;
+    } else {
+       thrust::tie(output_data[i], output_is_valid) =
+            get_new_value<T, replacement_has_nulls>(i, input_data,
+                                      values_to_replace_begin,
+                                      values_to_replace_end,
+                                      d_replacement_values,
+                                      replacement_valid);
     }
+
+    /* output valid counts calculations*/
+    if (input_has_nulls or replacement_has_nulls){
+
+      bitmask &= __ballot_sync(active_mask, output_is_valid);
+
+      __shared__ uint32_t valid_sum[warp_size];
+
+      // init shared memory for block valid counts
+      if(threadIdx.x < warp_size) valid_sum[threadIdx.x] = 0;
+      __syncthreads();
+
+      const int wid = threadIdx.x / warp_size;
+      const int lane = threadIdx.x % warp_size;
+
+      if(lane == 0 && bitmask != 0){
+        output_valid[(int)(i/warp_size)] = bitmask;
+        valid_sum[wid] = __popc(bitmask);
```

Comment:
If you change this `=` to `+=`, then you can move the below `__syncthreads()`, `sum_warps` and `atomicAdd` outside of the `while` loop, so it's only done once per kernel.  And you can move the initial `valid_sum` declaration and initialization and first `___syncthreads()` before the `while` loop so it's also only done once. Just make sure you put them all inside the `if has_nulls` checks.

Reply:
Yes, I have changed in latest commit. Please review...